### PR TITLE
Better printing

### PR DESCRIPTION
**Pull request by mitchphillipson**
Improve how objects are displayed to the user. To do this I've included the dependency PrettyTables.jl, a utility to easily make nice looking tables. Related to Issue #125 

As an example, the new printing looks like:

```
Sector: N
Description: Two dimensional
┌──────────┬─────────────┬───────────┬─────────────┬───────┐
│    Index │ Lower Bound │ Benchmark │ Upper Bound │ Fixed │
├──────────┼─────────────┼───────────┼─────────────┼───────┤
│ (:a, :c) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :c) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:a, :d) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :d) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:a, :e) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :e) │         0.0 │       1.0 │     1.0e101 │ false │
└──────────┴─────────────┴───────────┴─────────────┴───────┘
```

Currently have implemented for Sectors, Commodities, Consumers, Aux, and Parameters. I plan to do similar for the other structures (Production, Input, Output, etc.), but they'll look necessarily different.

### CODE REVIEW

**Comment by mitchphillipson** [comment]
The function `add!` returns either a `SectorRef` or a DenseAxisArray. I'm having an issue where it's hard to type-check a DenseAxisArray type. 

But I think there are a few compounding issues. For example, consider the following

```
M1 = MPSGE.Model()

N = add!(M1,Sector(:N, indices = ([:a,:b],[:c,:d,:e]), description = "Two dimensional"))
```
In this case, N is a DenseAxisArray. It seems that `N[:a,:c]` should return a ScalarSector, but this isn't the case. It returns a SectorRef. And it is not super obvious how to extract information from this. If I want the benchmark value of this entry, is this the easiest way?
```
M1._sectors[N[:a,:c].index].benchmark[N[:a,:c].subindex_names...]
```
I have a several suggestions and WLOG I'll consider only Sectors:

### First suggestion
Two types, ScalarSector and IndexedSector. IndexedSector is a container of ScalarSectors

```
mutable struct ScalarSector
    #name::Symbol #I don't think this needs to know it's own name, actually it may
    benchmark::Float64
    lower_bound::Float64
    upper_bound::Float64
    description::String
    fixed::Bool

    function ScalarSector(; description::AbstractString="", lower_bound::Float64=0.0, upper_bound=10e100, benchmark::Float64=1., fixed=false)
        return new(name, benchmark, lower_bound, upper_bound, description, fixed)
    end
end

mutable struct IndexedSector
    #name::Symbol
    indices::Any #We should know the type
    description::String
    data::DenseAxisArray


    function IndexedSector(#name::Symbol,
                           indices; 
                           description::AbstractString="",  
                           benchmark::Float64=1., 
                           lower_bound::Float64=0.0, 
                           upper_bound::Float64=10e100, 
                           fixed=false)

        return new(indices, description, DenseAxisArray(fill(ScalarSector(description,benchmark,lower_bound,upper_bound,fixed),length.(indices)),indices...)
    end
end
```

This eliminates the abstract type Sector, which you don't appear to be using outside of `model.jl`. However, this does currently duplicate the index set (it's also in the DenseAxisArray). After this define a getter to extract individual ScalarSectors from an IndexedSector. 

~~This would probably require a fairly substantial rewrite of code~~ I see odow has a pull request #124 removing eval from building variables. It's very easy to modify that code to make these changes, just modify `add_indexed_variable!`.

### Second Suggestion

Change the `_sectors` variable to be a dictionary rather than a list. The type signature would be `Symbol => Union{ScalarSector,IndexSector}` where the key is the name of the Sector. This is where I think you can eliminate the name from the structs. 

This would also allow accessing elements of the MPSGE model easier. As an unrefined example:

```
function Base.getindex(M::MPSGE.Model,key::Symbol)
    if key in keys(M._sectors)
        return M._sectors[key]
    elseif ...

    end
    error("Key $key not in MPSGE Model")
end
```

### Third Suggestion

Don't overwrite Model. Make the primary data structure be called MPSGE. Using this I will have several models running around, one to calibrate the data, one algebraic version of the model, and one MPSGE. I don't want to have to specify `JuMP.Model` and `MPSGE.Model`. 

I would recommend a function `build_model(M::MPSGE)` that returns a JuMP model. MPSGE could remember this model for later use.


This became much, much longer than I anticipated and I'm sure I'll have much, much more. I'm going to tag both of you so you see this. @davidanthoff @EliLazarus

File: src/model.jl [hunk 1]
```diff
@@ -991,3 +991,188 @@ function set_upper_bound(sector::SectorRef, u_bound::Float64)
...
 end
+
+
+
+#Parameter::ScalarParameter, ::IndexedParameter
+
+#Commodity::ScalarCommodity, ::IndexedCommodity
+#Sector::ScalarSector,       ::IndexedSector
+#Consumer::ScalarConsumer,   ::IndexedConsumer
+#Aux::ScalarAux,             ::IndexedAux
```
Comment (davidanthoff) [inline src/model.jl:1002]: This can be removed, right?

File: Project.toml [hunk 2]
```diff
@@ -6,17 +6,18 @@ version = "0.3.1-DEV"
 Complementarity = "a9b2a840-c9d5-5181-a245-8df664c2d6e7"
 JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
 MacroTools = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
-
-[extras]
-XLSX = "fdbf4ff8-1666-58a4-91e7-1b58723a45e0"
-TestItemRunner = "f8b46487-2199-4994-9208-9a1283c18c0a"
-Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+PrettyTables = "08abe8d2-0d0c-5749-adfa-8a2ac140af0d"
```
Comment (davidanthoff) [inline Project.toml:9]: Can you add `PrettyTables` to `[compat]` as well? We won't be able to register a new version without it. You can just make the current version of `PrettyTables` the min we need.

**Review by davidanthoff — CHANGES_REQUESTED** [review]
Very nice, we should definitely merge this.

Few change requests:
- I think we should maybe move all the `show` and printing methods into a separate `show.jl` file. `model.jl` is already large.
- I'm just waiting for a new JuMP.jl tag, and then I'll merge https://github.com/anthofflab/MPSGE.jl/pull/122. I want to merge #122 first, so I think it might be easiest if you either open this PR here against that branch, or maybe we just wait with this merge here until we've merged #122? Moving the show code into a separate file would probably also make the merge trivial (if there is any problem in the first place).

**Comment by davidanthoff** [comment]
Oh, and on the other points, should we move those to an issue or do a call or something? This PR seems focused and (almost) ready to be merged, so probably best to keep these things separate?

File: src/model.jl (hunk 1 above)
Comment (mitchphillipson) [reply src/model.jl:1002]: Yeah, absolutely. Reference for myself to ensure I hit all the points.

File: Project.toml (hunk 2 above)
Comment (mitchphillipson) [reply Project.toml:9]: Done

**Comment by mitchphillipson** [comment]
> Oh, and on the other points, should we move those to an issue or do a call or something? This PR seems focused and (almost) ready to be merged, so probably best to keep these things separate?

Yes, I think a separate issue is the correct call. I got to typing and kept having ideas.

**Comment by mitchphillipson** [comment]
There are a couple of issues here that are not great. Currently a DenseAxisArray of a SectorRef does not print well, since each item in the array is a SectorRef it'll print the entire table multiple times

```
M1 = MPSGE.Model()

N = add!(M1,Sector(:N, indices = ([:a,:b],[:c,:d,:e]), description = "Two dimensional"))

N
```
will print this table

```
Sector: N
Description: Two dimensional
┌──────────┬─────────────┬───────────┬─────────────┬───────┐
│    Index │ Lower Bound │ Benchmark │ Upper Bound │ Fixed │
├──────────┼─────────────┼───────────┼─────────────┼───────┤
│ (:a, :c) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :c) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:a, :d) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :d) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:a, :e) │         0.0 │       1.0 │     1.0e101 │ false │
│ (:b, :e) │         0.0 │       1.0 │     1.0e101 │ false │
└──────────┴─────────────┴───────────┴─────────────┴───────┘
```

6 times.

I also haven't incorporating nesting structures into Production, but do have a general print looking like MPSGE code from GAMS.

```
MPSGE model with 4 sectors, 5 commodities and 1 consumers.
  Sectors: X (bm=1.0), Y (bm=1.0), W (bm=1.0), G (bm=1-dimensional DenseAxisArray{Float64,1,...} with index sets:
    Dimension 1, [:a, :b]
And data, a 2-element Vector{Float64}:
 1.0
 1.0)
  Commodities: PX (bm=1.0), PY (bm=1.0), PL (bm=1.0), PK (bm=1.0), PW (bm=1.0)
  Consumers: CONS (bm=1.0)

Production: X	s: 1.0

O: PX	Q: 100.0	
I: PL	Q: 25.0	A: CONS	T: 0.0
I: PK	Q: 75.0	A: CONS	T: 0.0

Production: Y	

O: PY	Q: 100.0	
I: PL	Q: 75.0	
I: PK	Q: 25.0	

Production: W	

O: PW	Q: 200.0	
I: PX	Q: 100.0	
I: PY	Q: 100.0	

Demand: CONS

D: PW	Q: 200.0	
E: PL	Q: 100.0
E: PK	Q: 100.0
```

I still have some work to do in a few spots.

**Comment by mitchphillipson** [comment]
Question about the code to add a Sector to an MPSGEModel

```
function add!(m::Model, s::IndexedSector)
    m._jump_model = nothing
    push!(m._sectors, s)

    temp_array = Array{SectorRef}(undef, length.(s.indices)...)

    for i in CartesianIndices(temp_array)
        temp_array[i] = SectorRef(m, length(m._sectors), i, Tuple(s.indices[j][v] for (j,v) in enumerate(Tuple(i))))
    end
    return JuMP.Containers.DenseAxisArray(temp_array, s.indices...)
end
```

Is there a specific reason it returns a DenseAxisArray? It seems like it would make more sense to return s.

```
function add!(m::Model, s::IndexedSector)
    m._jump_model = nothing
    push!(m._sectors, s)
```

My guess is it's to make things like Production easier as the commodity is forced to be a CommodityRef.

This may, again, be better in Issues, but it's messing up how things are printed.

File: src/show.jl [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    #print(io,get_full(S))
+    if isnothing(S.subindex)
+        return print(io,get_full(S))
+    else
+        names = S.subindex_names
+        full_S = get_full(S)
+        return print(io,"$(names)\tbm: $(full_S.benchmark[names...])")
+    end
+
+end
```
Comment (mitchphillipson) [inline src/show.jl:53]: I've solve the issue of the dense axis arrays not printing correctly is a fairly hacky way. I check if the SectorRef refers to an IndexSector by testing if the subindex is nothing. It's not a great solution.

File: src/show.jl [hunk 4]
```diff
@@ -0,0 +1,224 @@
+function Base.show(io::IO, m::Model)
+    println(io, "MPSGE model with $(length(m._sectors)) sectors, $(length(m._commodities)) commodities and $(length(m._consumers)) consumers.")
+
+    if length(m._sectors) > 0
+        print(io, "  Sectors: ")
+        print(io, join(["$(s.name)" for s in m._sectors], ", "))
+        println(io)
+    end
+
```
Comment (mitchphillipson) [inline src/show.jl:9]: Removed benchmarks from printing when a model is printed. This caused over printing when an IndexedSector printed as an IndexedSector is potentially huge.

**Comment by mitchphillipson** [comment]
I believe everything prints as expected and this is ready for review.

**Comment by mitchphillipson** [comment]
Updated this to have no conflicts with the base branch. Issue was the new JuMP version.

**Comment by davidanthoff** [comment]
Fantastic, thanks so much!

**Pull request merged**